### ISSUE TITLE
java.lang.UnsupportedOperationException:

### ISSUE THREAD

**Issue by gigafone** [technomancy/leiningen#2091]
Can you help improve the following error message?

If I make the following mistake and try:

```sh
cd bar # note: there is no `project.clj` in here
lein run -m foo
```

I get:

```sh
clojure.lang.Compiler$CompilerException: java.lang.UnsupportedOperationException: Unknown Collection type, compiling:(NO_SOURCE_PATH:0:0)
 at clojure.lang.Compiler.analyzeSeq (Compiler.java:6875)
    clojure.lang.Compiler.analyze (Compiler.java:6669)
    clojure.lang.Compiler.eval (Compiler.java:6924)
    clojure.lang.Compiler.eval (Compiler.java:6917)
    clojure.lang.Compiler.eval (Compiler.java:6890)
    clojure.core$eval.invokeStatic (core.clj:3105)
    clojure.core$eval.invoke (core.clj:3101)
    leiningen.core.eval$fn__5670.invokeStatic (eval.clj:349)
    leiningen.core.eval/fn (eval.clj:338)
    clojure.lang.MultiFn.invoke (MultiFn.java:233)
    leiningen.core.eval$eval_in_project.invokeStatic (eval.clj:373)
    leiningen.core.eval$eval_in_project.invoke (eval.clj:363)
    leiningen.core.eval$eval_in_project.invokeStatic (eval.clj:367)
    leiningen.core.eval$eval_in_project.invoke (eval.clj:363)
    leiningen.run$run_main.invokeStatic (run.clj:128)
    leiningen.run$run_main.invoke (run.clj:123)
    leiningen.run$run.invokeStatic (run.clj:155)
    leiningen.run$run.doInvoke (run.clj:132)
    clojure.lang.RestFn.invoke (RestFn.java:439)
    clojure.lang.Var.invoke (Var.java:388)
    clojure.lang.AFn.applyToHelper (AFn.java:160)
    clojure.lang.Var.applyTo (Var.java:700)
    clojure.core$apply.invokeStatic (core.clj:648)
    clojure.core$apply.invoke (core.clj:641)
    leiningen.core.main$partial_task$fn__5829.doInvoke (main.clj:272)
    clojure.lang.RestFn.applyTo (RestFn.java:139)
    clojure.lang.AFunction$1.doInvoke (AFunction.java:29)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.core$apply.invokeStatic (core.clj:648)
    clojure.core$apply.invoke (core.clj:641)
    leiningen.core.main$apply_task.invokeStatic (main.clj:322)
    leiningen.core.main$apply_task.invoke (main.clj:308)
    leiningen.core.main$resolve_and_apply.invokeStatic (main.clj:328)
    leiningen.core.main$resolve_and_apply.invoke (main.clj:324)
    leiningen.core.main$_main$fn__5895.invoke (main.clj:401)
    leiningen.core.main$_main.invokeStatic (main.clj:394)
    leiningen.core.main$_main.doInvoke (main.clj:391)
    clojure.lang.RestFn.invoke (RestFn.java:436)
    clojure.lang.Var.invoke (Var.java:388)
    clojure.lang.AFn.applyToHelper (AFn.java:160)
    clojure.lang.Var.applyTo (Var.java:700)
    clojure.core$apply.invokeStatic (core.clj:646)
    clojure.main$main_opt.invokeStatic (main.clj:314)
    clojure.main$main_opt.invoke (main.clj:310)
    clojure.main$main.invokeStatic (main.clj:421)
    clojure.main$main.doInvoke (main.clj:384)
    clojure.lang.RestFn.invoke (RestFn.java:482)
    clojure.lang.Var.invoke (Var.java:401)
    clojure.lang.AFn.applyToHelper (AFn.java:171)
    clojure.lang.Var.applyTo (Var.java:700)
    clojure.main.main (main.java:37)
Caused by: java.lang.UnsupportedOperationException: Unknown Collection type
 at clojure.lang.Compiler$EmptyExpr.emit (Compiler.java:2951)
    clojure.lang.Compiler$InvokeExpr.emitArgsAndCall (Compiler.java:3732)
    clojure.lang.Compiler$InvokeExpr.emit (Compiler.java:3675)
    clojure.lang.Compiler$InvokeExpr.emitArgsAndCall (Compiler.java:3732)
    clojure.lang.Compiler$InvokeExpr.emit (Compiler.java:3675)
    clojure.lang.Compiler$BodyExpr.emit (Compiler.java:6042)
    clojure.lang.Compiler$LetExpr.doEmit (Compiler.java:6399)
    clojure.lang.Compiler$LetExpr.emit (Compiler.java:6352)
    clojure.lang.Compiler$BodyExpr.emit (Compiler.java:6045)
    clojure.lang.Compiler$ObjMethod.emitBody (Compiler.java:5764)
    clojure.lang.Compiler$FnMethod.doEmitStatic (Compiler.java:5428)
    clojure.lang.Compiler$FnMethod.emit (Compiler.java:5393)
    clojure.lang.Compiler$FnExpr.emitMethods (Compiler.java:3887)
    clojure.lang.Compiler$ObjExpr.compile (Compiler.java:4497)
    clojure.lang.Compiler$FnExpr.parse (Compiler.java:4049)
    clojure.lang.Compiler.analyzeSeq (Compiler.java:6866)
    clojure.lang.Compiler.analyze (Compiler.java:6669)
    clojure.lang.Compiler.eval (Compiler.java:6924)
    clojure.lang.Compiler.eval (Compiler.java:6917)
    clojure.lang.Compiler.eval (Compiler.java:6890)
    clojure.core$eval.invokeStatic (core.clj:3105)
    clojure.core$eval.invoke (core.clj:3101)
    leiningen.core.eval$fn__5670.invokeStatic (eval.clj:349)
    leiningen.core.eval/fn (eval.clj:338)
    clojure.lang.MultiFn.invoke (MultiFn.java:233)
    leiningen.core.eval$eval_in_project.invokeStatic (eval.clj:373)
    leiningen.core.eval$eval_in_project.invoke (eval.clj:363)
    leiningen.core.eval$eval_in_project.invokeStatic (eval.clj:367)
    leiningen.core.eval$eval_in_project.invoke (eval.clj:363)
    leiningen.run$run_main.invokeStatic (run.clj:128)
    leiningen.run$run_main.invoke (run.clj:123)
    leiningen.run$run.invokeStatic (run.clj:155)
    leiningen.run$run.doInvoke (run.clj:132)
    clojure.lang.RestFn.invoke (RestFn.java:439)
    clojure.lang.Var.invoke (Var.java:388)
    clojure.lang.AFn.applyToHelper (AFn.java:160)
    clojure.lang.Var.applyTo (Var.java:700)
    clojure.core$apply.invokeStatic (core.clj:648)
    clojure.core$apply.invoke (core.clj:641)
    leiningen.core.main$partial_task$fn__5829.doInvoke (main.clj:272)
    clojure.lang.RestFn.applyTo (RestFn.java:139)
    clojure.lang.AFunction$1.doInvoke (AFunction.java:29)
    clojure.lang.RestFn.applyTo (RestFn.java:137)
    clojure.core$apply.invokeStatic (core.clj:648)
    clojure.core$apply.invoke (core.clj:641)
    leiningen.core.main$apply_task.invokeStatic (main.clj:322)
    leiningen.core.main$apply_task.invoke (main.clj:308)
    leiningen.core.main$resolve_and_apply.invokeStatic (main.clj:328)
    leiningen.core.main$resolve_and_apply.invoke (main.clj:324)
    leiningen.core.main$_main$fn__5895.invoke (main.clj:401)
    leiningen.core.main$_main.invokeStatic (main.clj:394)
    leiningen.core.main$_main.doInvoke (main.clj:391)
    clojure.lang.RestFn.invoke (RestFn.java:436)
    clojure.lang.Var.invoke (Var.java:388)
    clojure.lang.AFn.applyToHelper (AFn.java:160)
    clojure.lang.Var.applyTo (Var.java:700)
    clojure.core$apply.invokeStatic (core.clj:646)
    clojure.main$main_opt.invokeStatic (main.clj:314)
    clojure.main$main_opt.invoke (main.clj:310)
    clojure.main$main.invokeStatic (main.clj:421)
    clojure.main$main.doInvoke (main.clj:384)
    clojure.lang.RestFn.invoke (RestFn.java:482)
    clojure.lang.Var.invoke (Var.java:401)
    clojure.lang.AFn.applyToHelper (AFn.java:171)
    clojure.lang.Var.applyTo (Var.java:700)
    clojure.main.main (main.java:37)
```

Can perhaps leiningen instead give a message along the lines of one of these?

* project.clj not found
* namespace not found: foo

I'm running leiningen 2.6.1

**Comment by gigafone** [comment]
A friend told me that he got a reasonable error message on Leiningen version 2.5.3, along the lines of:

```
Can't find 'foo' as .class or .clj for lein run: please check the spelling.
```

**Comment by hyPiRion** [comment]
Woah, yeah, that's just a terrible error message to print out.

If this was better in 2.5.3 then this is a regression, and I believe it shouldn't be too hard to fix this. Perhaps that's even a newbie task.

It may be that the error message came from another place though, in which case it may or may not be that easy to fix it.

**Comment by gigafone** [comment]
It might be the worst I've seen, actually. :)

I changed my `~/.lein/profiles.clj` file to be simply `{}` to rule out any relationship to it and got the same error:

```
clojure.lang.Compiler$CompilerException: java.lang.UnsupportedOperationException: Unknown Collection type, compiling:(NO_SOURCE_PATH:0:0)
 at clojure.lang.Compiler.analyzeSeq (Compiler.java:6875)
    clojure.lang.Compiler.analyze (Compiler.java:6669)
...
```

**Comment by 3duard0** [comment]
Added PR to fix this issue, more details about the issue on the PR.

**Comment by hyPiRion** [comment]
Fixed by #2157.
